Timing and training score in *SearchCV.results_

eyc88 opened the pull request:
<!--
Thanks for contributing a pull request! Please ensure you have taken a look at
the contribution guidelines: https://github.com/scikit-learn/scikit-learn/blob/master/CONTRIBUTING.md#Contributing-Pull-Requests
-->
#### Reference Issue

<!-- Example: Fixes #1234 -->

Resolves #6894 Timing in *SearchCV.results_
               #6895 Training Score in *SearchCV.results_
#### What does this implement/fix? Explain your changes.

The _fit method in BaseSearchCV calls the _fit_and_score function from module _validation.
_fit_and_score has the ability to report train_score, but not by default.  On the other hand,
the __init__ of BaseSearchCV does not accept return_train_score as an optional keyword.  As a result, there was no way to have the results_ attribute receving train score.

The fix is to add the keyword return_train_score to the __init__ of BaseSearchCV (which is inherited by both GridSearchCV and RandomizedSearchCV).  Default is still set to False.  When return_train_score is set to True, a set of addition steps is taken to put train_score into results_

The timing information is always available but was not written into "results_".  This patch also writes timing info into results_ by default.
#### Any other comments?

Test is also re-written.  In the test I explicitly choose return_train_score=True for better coverage.

<!--
Please be aware that we are a loose team of volunteers so patience is
necessary; assistance handling other issues is very welcome. We value
all user contributions, no matter how minor they are. If we are slow to
review, either the pull request needs some benchmarking, tinkering,
convincing, etc. or more likely the reviewers are simply busy. In either
case, we ask for your understanding during the review process.
For more information, see our FAQ on this topic:
http://scikit-learn.org/dev/faq.html#why-is-my-pull-request-not-getting-any-attention.

Thanks for contributing!
-->

Commit:
Resolved issue #6894 and #6895:
    Now *SearchCV.results_ includes both timing and training scores.

Commit:
wrote new test (sklearn/model_selection/test_search.py)
and new doctest (sklearn/model_selection/_search.py)

Commit:
added a few more lines in the docstring of GridSearchCV and RandomizedSearchCV.

jnothman commented on sklearn/model_selection/_search.py:
```diff
@@ -822,13 +822,17 @@ class GridSearchCV(BaseSearchCV):
         NOTE that the key ``'params'`` is used to store a list of parameter
-        settings dict for all the parameter candidates.
+        settings dict for all the parameter candidates.  Besides,
+        'train_mean_score', 'train_split*_score', ... will be present when
```
Please follow the style of using double backticks to indicate code.

raghavrv replied:
Just noting that it should be backticks surrounding the single quoted key.

\`\`'train_mean_score'\`\` (which will display as `'train_mean_score'`)

jnothman commented on sklearn/model_selection/_search.py:
```diff
@@ -1013,13 +1060,17 @@ class RandomizedSearchCV(BaseSearchCV):
         NOTE that the key ``'params'`` is used to store a list of parameter
-        settings dict for all the parameter candidates.
+        settings dict for all the parameter candidates.  Besides,
+        'train_mean_score', 'train_split*_score', ... will be present when
```
Place this under the `return_train_score` description.

jnothman commented on sklearn/model_selection/_search.py:
```diff
@@ -552,17 +565,39 @@ def _fit(self, X, y, labels, parameter_iterable):
-        stds = np.sqrt(np.average((test_scores - means[:, np.newaxis]) ** 2,
-                                  axis=1, weights=weights))
+
+        time = np.array(_, dtype=np.float64).reshape(n_candidates, n_splits)
```
No, we don't use `_` as a variable name. In the former code, it means "we don't want to reuse this variable".

jnothman commented on sklearn/model_selection/_search.py:
```diff
@@ -552,17 +565,39 @@ def _fit(self, X, y, labels, parameter_iterable):
+        time = np.array(_, dtype=np.float64).reshape(n_candidates, n_splits)
+        time_means = np.average(time, axis=1, weights=weights)
+        time_stds =  np.sqrt(
+                np.average((time - time_means[:, np.newaxis]) ** 2,
```
Please use PEP8 indentation.

raghavrv commented:
Thanks for the PR!!

jnothman commented on sklearn/model_selection/_search.py:
```diff
@@ -552,17 +565,39 @@ def _fit(self, X, y, labels, parameter_iterable):
+                np.average((time - time_means[:, np.newaxis]) ** 2,
+                           axis=1, weights=weights))
+        if self.return_train_score:
+            train_means = np.average(train_scores, axis=1, weights=weights)
```
It doesn't make sense to weight the train scores by the number of test samples. Perhaps we should weight by the number of training samples when `iid=True`, or report an unweighted average. For times, weighting by training instances is more appropriate than by test, but leaving it unweighted might be best.

amueller replied:
I agree. Also, I think ``iid`` is a bad hack and we need to fix the scorers to accept both training and test data to solve this properly :-/

jnothman replied:
Sorry if I'm a bit slow here, but please elucidate at some point. What's `iid` trying to hack? Why does accepting training data in the scorer help?

amueller replied:
It's trying to hack the R^2, IIRC. @agramfort can say more about this, I think.
The way we compute R^2 is weird because we are using the test set mean, not the training set mean. Though actually after thinking about it more, maybe accepting training data is too big a change just for that. I thought it would also resolve some unsupervised issued, but maybe it doesn't.

jnothman commented:
Perhaps we should avoid code duplication by defining `_result_scores` and reusing it for train and test, e.g. `self.results_.update(_result_scores('test_', test_scores, weights))`.

jnothman commented on sklearn/model_selection/_search.py:
```diff
@@ -1104,4 +1155,4 @@ def fit(self, X, y=None, labels=None):
                                           self.n_iter,
                                           random_state=self.random_state)
-        return self._fit(X, y, labels, sampled_params)
+        return self._fit(X, y, labels, sampled_params)
```
You have strangely removed the newline character from the end of this file.

jnothman commented on sklearn/model_selection/_search.py:
```diff
@@ -1074,15 +1125,15 @@ class RandomizedSearchCV(BaseSearchCV):
-            estimator=estimator, scoring=scoring, fit_params=fit_params,
-            n_jobs=n_jobs, iid=iid, refit=refit, cv=cv, verbose=verbose,
-            pre_dispatch=pre_dispatch, error_score=error_score)
+              estimator=estimator, scoring=scoring, fit_params=fit_params,
```
This is not PEP8-compliant indentation.

jnothman commented:
I've renamed the PR to something clearer.

amueller commented on sklearn/model_selection/_search.py:
```diff
@@ -371,7 +371,7 @@ class BaseSearchCV(six.with_metaclass(ABCMeta, BaseEstimator,
                  fit_params=None, n_jobs=1, iid=True,
                  refit=True, cv=None, verbose=0, pre_dispatch='2*n_jobs',
-                 error_score='raise'):
+                 error_score='raise', return_train_score=False):
```
I know it was in the issue, but I'm not sure why we want to make that optional / not the default. Computing the training scores is really cheap compared to everything else that goes on. There are already a lot of parameters, not sure if this one is particularly helpful. @jnothman ?

amueller replied:
I guess making it optional is helpful if you have a very large training set on which evaluation takes a long time because of an very expensive evaluation procedure. But I think we should default to ``True`` both for convenience and discoverability.

Commit:
Revised code according to suggestions.

amueller commented:
It would be nice to add the example of #1742 but I think that's not that important. LGTM.

amueller commented:
wait, sorry I thought this was only the timing. For the training scores, we should really add an example and tests.

amueller commented:
@eyc88 I think you said you might not have time to work on this now. Should someone else take over?

eyc88 commented:
@amueller If you just want to test that the training score is smaller or equal to 1, I can deliver it soon (one thing I would like to respond to @ogrisel is that training score can be exactly 1).  I am now travelling so there could be some delay.  All that being said, I don't mind if someone wants to take it over.

Commit:
Add a few more lines to test_grid_search_results():
    1. check test_rank_score always >= 1
    2. check all regular scores (test/train_mean/std_score) and timing >= 0
    3. check all regular scores <= 1
Note that timing can be greater than 1 in general, and std of regular scores
always <= 1 because the scores are bounded between 0 and 1.

eyc88 commented:
@amueller Check this out.  Let me know if this is what you had in mind.  We can discuss.

jnothman commented:
A test for training score would usually require a more explicit test that we're reporting the right number. We want, for instance, the tests to be fairly robust to this all being reimplemented.

amueller commented:
Yeah so a good test would be to instantiate a cross-validation object and do the calculations "by hand" for some simple example and check that ``GridSearchCV`` does them correctly.

eyc88 commented:
Sounds good.  I will take a closer look and write a new version of test once I got some free time.

eyc88 commented:
@jnothman When are you guys rolling out 0.18?  I can sqeeze some bandwidth if the deadline is approaching...

jnothman commented:
Within the next week or two.

amueller commented:
I think tests were my only issue. and I'd love to see this in 0.18

raghavrv commented:
@eyc88 Do you mind if I take this over?

eyc88 commented:
@raghavrv No.

lesteve commented:
Closing in favour of #7325.
